core/src/main/scala/org/apache/spark/SparkHadoopConf.scala

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark
+
+import java.net.URL
+
+import org.apache.hadoop.conf.Configuration
+
+import org.apache.spark.deploy.SparkHadoopUtil
+
+/**
+ * Hadoop Configuration for the Hadoop code (e.g. file systems).
+ */
+
+class SparkHadoopConf(conf: Configuration) {
+
+  def get: Configuration = conf
+
+  def get(name: String): String = conf.get(name)
+
+  def set(conf: Configuration, entries: (String, String)*): Unit = {
+    entries.foreach { case (name, value) =>
+      conf.set(name, value)
+    }
+  }
+
+  def set(entries: (String, String)*): Unit = {
+    set(conf, entries: _*)
+  }
+
+  def set(name: String, value: String): Unit = set(conf, name -> value)
+
+  def unset(name: String): Unit = conf.unset(name)
+
+  def addResource(url: URL): Unit = conf.addResource(url)
+}
+
+object SparkHadoopConf {
+  private var hadoopConf: ThreadLocal[SparkHadoopConf] = _
+
+  def init(conf: SparkConf): Unit = {
+    hadoopConf = new ThreadLocal[SparkHadoopConf]() {
+      override def initialValue: SparkHadoopConf = {
+        new SparkHadoopConf(SparkHadoopUtil.get.newConfiguration(conf))
+      }
+    }
+  }
+
+  def get(): SparkHadoopConf = hadoopConf.get()
```

Comment:
This May return null, perhaps make it return a `Option[SparkHadoopConf]`

Reply:
We'll always get an initial value by invoking `initialValue()` when calling `get()` on thread local variable. So, when does it return null ?

Reply:
It's never documented any where that we MUST call `initialValue` before using `get()`, also it's not transactional, so it will almost certainly be broken in the future.

Reply:
I think JDK guarantees this ? Please see the JDK doc above `initialValue`:

```
     * Returns the current thread's "initial value" for this
     * thread-local variable.  This method will be invoked the first
     * time a thread accesses the variable with the {@link #get}
     * method, unless the thread previously invoked the {@link #set}
     * method, in which case the {@code initialValue} method will not
     * be invoked for the thread.  Normally, this method is invoked at
     * most once per thread, but it may be invoked again in case of
     * subsequent invocations of {@link #remove} followed by {@link #get}.
     *
     * <p>This implementation simply returns {@code null}; if the
     * programmer desires thread-local variables to have an initial
     * value other than {@code null}, {@code ThreadLocal} must be
     * subclassed, and this method overridden.  Typically, an
     * anonymous inner class will be used.
```